cabby/geo/walk.py

```diff
@@ -38,56 +52,304 @@ def compute_route(start_point: Point, end_point: Point, graph: nx.MultiDiGraph,
+
+
+def get_single_sample(map: map_structure.Map):
+    '''Sample start and end point, a pivot landmark and route and save to file.
+    Arguments:
+      map: The map of a specific region.
+    Returns:
+      A start and end point, a pivot landmark and route.
+    '''
+    result = get_points_and_route(map)
+    if result is None:
+        return
+    end_point, start_point, route, main_pivot, near_pivot = result
+    instruction = "Starting at {0} walk past {1} and your goal is {2}, near {3}.".format(
+        start_point['name'], main_pivot['main_tag'], end_point['name'], near_pivot['main_tag'])
+
+    main_pivot['centroid'] = main_pivot['geometry'] if isinstance(
+        main_pivot['geometry'], Point) else main_pivot['geometry'].centroid
+
+    near_pivot['centroid'] = near_pivot['geometry'] if isinstance(
+        near_pivot['geometry'], Point) else near_pivot['geometry'].centroid
+
+    gdf = gpd.GeoDataFrame({'end': end_point['name'], 'start': start_point['name'], 'main_pivot': main_pivot['main_tag'],
+                            'near_pivot': near_pivot['main_tag'], 'instruction': instruction}, index=[0])
+
+    gdf['geometry'] = start_point['centroid']
+    gdf_end = gpd.GeoDataFrame(geometry=[end_point['centroid']])
+    gdf_main_pivot = gpd.GeoDataFrame(geometry=[main_pivot['centroid']])
+    gdf_near_pivot = gpd.GeoDataFrame(geometry=[near_pivot['centroid']])
+    gdf_route = gpd.GeoDataFrame(
+        geometry=[Polygon(route['geometry'].tolist())])
+    return gdf, gdf_end, gdf_main_pivot, gdf_near_pivot, gdf_route
+
+
+def get_samples(path: Text, map: map_structure.Map, n_samples: int):
+    '''Sample start and end point, a pivot landmark and route and save to file.
+    Arguments:
+      path: The path to which the data will be appended.
+      map: The map of a specific region.
+      n_samples: the max number of samples to generate.
+    '''
+    gdf_start_list = gpd.GeoDataFrame(
+        columns=["start", "end", "main_pivot", "near_pivot", "instruction", "geometry"])
+
+    gdf_end_list = gpd.GeoDataFrame(columns=["geometry"])
+    gdf_route_list = gpd.GeoDataFrame(columns=["geometry"])
+    gdf_main_list = gpd.GeoDataFrame(columns=["geometry"])
+    gdf_near_list = gpd.GeoDataFrame(columns=["geometry"])
 
-  # Select pivot.
+    for i in range(n_samples):
+        result = get_single_sample(map)
+        if result is None:
+            continue
+        gdf_start, gdf_end, gdf_main_pivot, gdf_near_pivot, gdf_route = result
+        gdf_start_list = gdf_start_list.append(gdf_start, ignore_index=True)
+        gdf_end_list = gdf_end_list.append(gdf_end, ignore_index=True)
+        gdf_route_list = gdf_route_list.append(gdf_route, ignore_index=True)
+        gdf_main_list = gdf_main_list.append(gdf_main_pivot, ignore_index=True)
+        gdf_near_list = gdf_near_list.append(gdf_near_pivot, ignore_index=True)
 
-  pass
+    gdf_start_list.to_file(path, layer='start', driver="GPKG")
```

Comment:
Do we need to know the driver name when we are loading these? If so (and maybe regardless of that), we should have it in a variable, e.g. _Geo_DataFrame_Driver (or something like that).